brave/brave-ios
# Fix #3228: Implement Password Sync

## Summary of Changes

This Pull Request contains 

- Adding Passwords/Logins to Sync Profile
- History Fetch/Add/Delete with Core and Sync
- Passwords/Logins local items Migration

###  Passwords/Logins Migration:

We are adding password brave-core migration to our existing migrator used for bookmarks/history and making all 3 sync types work together and handle error cases related for both of the types.

### Passwords Fetch/Add/Delete/Update with Core and Sync

We updated LoginsHelper class to work with new passwordAPI from brave-core, added helper methods for update/save and autofill the credentials. Trigger add or update functionality according to item existence. Additionally the the check for detecting if frame is nor modified when main frame is not used is updated and test for this security piece is added.

In addition the functionality in LoginsList and LoginsInfo controller are updated to use new password API. Listeners are added to info screen to update the list content when data is updated in another sync'ed device.

### Sync Profile

Enable/Disable Passwords Sync Types is added. Sync Settings will contain a new item which will allow user to toggle Password Sync type according to users preference. Items will not start syncing until the item(toggle) is turned on.

This pull request fixes #3228

Brave Core PR is merged and uplifted to 1.37.x and security review is completed for the core side.

Core PR For Password Sync: https://github.com/brave/brave-core/pull/12360

Sec Review: https://github.com/brave/security/issues/765

## Submitter Checklist:


## Test Plan:

Test Migration: The test cases and conditions are same with other sync 2 sync types (bookmarks-history). Tries migration 3 times but for all 3 types together in this case.

- Update application with existing passwords
- Check migration is happened and data is same

Normal Add / Update

- Fresh install or updated app
- Navigate to a website like [wikipedia.com](https://en.m.wikipedia.org/w/index.php?title=Special:UserLogin&returnto=Main+Page)
- Enter Credentials and press login
- Check Toast appears from bottom and click save
- Go Settings -> Logins & Passwords and check login entry is saved

![password1](https://user-images.githubusercontent.com/6643505/156453511-52495e26-3e7d-4077-bf1f-240ad325ce49.png)

Update Scenario can be tested over an auto-filled credential with a different password.

Test Sync: Same testing cases with bookmarks and history release and  joining to sync user can turn on/off which type of sync should be enabled including password

- Can test turning on different sync types and check If their functionality can be turned on/off

The functionality of Logins & Passwords screen can be seen https://github.com/brave/brave-ios/issues/4112 and these scenarios can be used with sync'ed entries.

## Screenshots:

![151589108-efb1e6b2-cd33-4487-9531-44d224defb52](https://user-images.githubusercontent.com/6643505/156454520-767de38a-7239-4e1d-8678-fc2bfceb470d.png)

![151595068-22ec3f70-9cd0-47da-93de-00cbc530a739](https://user-images.githubusercontent.com/6643505/156454565-d4ee684f-cd9e-4254-8fbb-09921743d3d7.png)

![password2 21 1](https://user-images.githubusercontent.com/6643505/156454849-5ccc8303-ae4b-46be-bce3-a934e2b9d14e.png)


## Reviewer Checklist:

  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
